### PR TITLE
[RLlib] CNN Models for RLModules

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
## Why are these changes needed?

The new Catalog needs another CNN implementation. This PR adds it.
This PR builds on the previous Model PR https://github.com/ray-project/ray/pull/32453

### CODE REVIEW

**Comment by ArturNiederfahrenhorst** [comment]
@kouroshHakha I'd like to point out that, due to the Encoder/Head division + specs, we can't produce the exact same standard models with these CNNs. I'll go over that in https://github.com/ray-project/ray/pull/32650 and point out why there. It's not critical though and the same models _can_ be created. Just not with default arguments.

File: rllib/core/models/configs.py [hunk 1]
```diff
@@ -83,6 +83,7 @@ class MLPHeadConfig(ModelConfig):
     hidden_layer_dims: List[int] = field(default_factory=lambda: [256, 256])
     hidden_layer_activation: str = "relu"
     output_activation: str = "linear"
+    output_dim: int = None
```
Comment (kouroshHakha) [inline rllib/core/models/configs.py:86]: docstring for this parameter?

File: rllib/core/models/configs.py [hunk 2]
```diff
@@ -107,6 +108,51 @@ def build(self, framework: str = "torch") -> Model:
...
 
+@ExperimentalAPI
+@dataclass
+class CNNEncoderConfig(ModelConfig):
+    """Configuration for a convolutional network.
+
+    See ModelConfig for usage details.
+
+    Attributes:
+        input_dims: The input dimension of the network. It cannot be None.
```
Comment (kouroshHakha) [inline rllib/core/models/configs.py:119]: What does input_dims mean for CNN? is it C, H, W? then it should be a tuple of three ints? not just one int? I am confused about this.

File: rllib/core/models/configs.py [hunk 3]
```diff
@@ -107,6 +108,51 @@ def build(self, framework: str = "torch") -> Model:
...
+        filter_specifiers: A list of lists, where each element of an inner list
+            contains elements of the form
+            `[number of filters, [kernel width, kernel height], stride]` to specify a
```
Comment (kouroshHakha) [inline rllib/core/models/configs.py:122]: number of filters is also known as  number of channels. might be worth including both.

File: rllib/core/models/configs.py [hunk 4]
```diff
@@ -107,6 +108,51 @@ def build(self, framework: str = "torch") -> Model:
             return TfMLPHead(self)
 
 
+@ExperimentalAPI
+@dataclass
+class CNNEncoderConfig(ModelConfig):
+    """Configuration for a convolutional network.
+
+    See ModelConfig for usage details.
```
Comment (kouroshHakha) [inline rllib/core/models/configs.py:116]: I think it's worth giving a concrete CNN example and explaining the stack (e.g. DQN paper architecture)
For example How would you describe this chain (where input images are represented as a 4x84x84 volume of 4 stacked grayscale images)
```
Conv2d(in_channels=4, out_channels=32, kernel_size=8, stride=4)
ReLU()
Conv2d(in_channels=32, out_channels=64, kernel_size=4, stride=2)
ReLU()
Conv2d(in_channels=64, out_channels=64, kernel_size=3, stride=1)
ReLU()
Flatten(),
Linear(in_features=3136, out_features=512)
```

File: rllib/core/models/configs.py [hunk 5]
```diff
@@ -107,6 +108,51 @@ def build(self, framework: str = "torch") -> Model:
...
+    Attributes:
+        input_dims: The input dimension of the network. It cannot be None.
+        filter_specifiers: A list of lists, where each element of an inner list
+            contains elements of the form
+            `[number of filters, [kernel width, kernel height], stride]` to specify a
+            convolutional layer stacked in order of the outer list.
+        filter_layer_activation: The activation function to use after each layer (
+            except for the output).
+        output_activation: The activation function to use for the output layer.
+        output_dim: The output dimension are [BATCH, output_dim]. We append a final
```
Comment (kouroshHakha) [inline rllib/core/models/configs.py:127]: I am confused about mentioning BATCH here all of a sudden. All of this dims have an implicit batch in them, right? Do you want to say that in the end we will add a flatten layer to obtain output_dim features?

File: rllib/core/models/torch/primitives.py [hunk 6]
```diff
@@ -56,3 +62,106 @@ def __init__(
 
     def forward(self, x):
         return self.mlp(x)
+
+
+class TorchCNN(nn.Module):
```
Comment (kouroshHakha) [inline rllib/core/models/torch/primitives.py:67]: I think structure wise, TorchCNN primitive should just be a stack of Conv2D layers + activations, but TorchCNNEncoder should do the flattening in the end as well. Right now first the primitive is not usable stand-alone if I don't want to flattening portion. This will also clearly separate responsibilities between primitives and encoders i.e. Encoder can be collection of primitives working together. It's not just simple pass-through to one of the primitives.

File: rllib/core/models/torch/primitives.py [hunk 7]
```diff
@@ -56,3 +62,106 @@ def __init__(
...
+        core_layers = []
+
+        # Add user-specified hidden convolutional layers first
+        width, height, in_depth = input_dims
+        in_size = [width, height]
+        for out_depth, kernel, stride in filter_specifiers:
+            padding, out_size = same_padding(in_size, kernel, stride)
+            # TODO(Artur): Inline SlimConv2d
+            core_layers.append(
+                SlimConv2d(
```
Comment (kouroshHakha) [inline rllib/core/models/torch/primitives.py:116]: Why do we want to use these layers, instead of the built-int Conv2D? What are we adding on top of it? Is it taking away the complexity of Conv2D? if not I would argue for removing one extra layer and directly link to Conv2D here.

File: rllib/core/models/torch/primitives.py [hunk 8]
```diff
@@ -56,3 +62,106 @@ def __init__(
...
+
+        layers = []
+        # Create some CNN layers
+        core_layers = []
+
+        # Add user-specified hidden convolutional layers first
+        width, height, in_depth = input_dims
+        in_size = [width, height]
+        for out_depth, kernel, stride in filter_specifiers:
+            padding, out_size = same_padding(in_size, kernel, stride)
```
Comment (kouroshHakha) [inline rllib/core/models/torch/primitives.py:113]: Also what does this function do exactly? Is this always needed? should this be enabled / disabled with an external flag?

File: rllib/core/models/torch/primitives.py [hunk 9]
```diff
@@ -56,3 +62,106 @@ def __init__(
...
+        layers.append(nn.Linear(int(width) * int(height), output_dim))
+        if output_activation is not None:
+            layers.append(output_activation())
+
+        # Create the cnn that potentially includes a flattened layer
+        self.cnn = nn.Sequential(*layers)
+
+    def forward(self, x):
+        # Permute b/c data comes in as [B, dim, dim, channels]:
+        inputs = x.permute(0, 3, 1, 2)
```
Comment (kouroshHakha) [inline rllib/core/models/torch/primitives.py:166]: This is why spec checking is required :) How would I know if the input is in the correct format. NICE !

File: rllib/core/models/torch/encoder.py [hunk 10]
```diff
@@ -65,6 +65,54 @@ def _forward(self, inputs: NestedDict, **kwargs) -> NestedDict:
...
+            filter_layer_activation=config.filter_layer_activation,
+            output_activation=config.output_activation,
+            output_dim=config.output_dim,
+        )
+
+    @override(Model)
+    def get_input_spec(self) -> Union[Spec, None]:
+        return SpecDict(
+            {
+                SampleBatch.OBS: TorchTensorSpec(
```
Comment (kouroshHakha) [inline rllib/core/models/torch/encoder.py:87]: Oh i love this 🤩

Comment (kouroshHakha) [reply rllib/core/models/torch/encoder.py:87]: This is much more elegant than doing if-else statements in forward function.

**Review by kouroshHakha — COMMENTED** [review]
Nice PR. Left a couple of comments.

File: rllib/core/models/configs.py (hunk 5 above)
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/configs.py:127]: Yeah, I removed the mentioning of BATCH. Thanks!

File: rllib/core/models/torch/primitives.py (hunk 6 above)
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/torch/primitives.py:67]: Yeah my idea was to reuse this Conv2d class we had before to be close to what we had before and after seeing it work, simply inline that code. I followed the second part of your advice about the primitive CNN including only the CNN itself. Let's keep the Conv2d CNN just for now to share more code with the legacy models. I'll add a todo to simply inline that code and remove unused bits once the transition has been made, ok?

File: rllib/core/models/torch/primitives.py (hunk 7 above)
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/torch/primitives.py:116]: Just like further up (https://github.com/ray-project/ray/pull/32515#discussion_r1113616936), I did it simply to share more code with the old Models until these are gone. Just to make it easier to not deviate from their behaviour!

File: rllib/core/models/torch/primitives.py (hunk 8 above)
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/torch/primitives.py:113]: This makes it so that the Torch and TF models are the same. Torch has no _SAME_ padding mode.

File: rllib/core/models/configs.py [hunk 11]
```diff
@@ -69,20 +69,45 @@ def _convert_to_lower_case_if_tf(string: str, framework: str) -> str:
...
+        output_dim = 2
+        output_activation = "linear"
+
+        Resulting stack in pseudocode:
+        Linear(4, 8)
+        ReLU()
+        Linear(8, 8)
+        ReLU()
+        Linear(8, 2)
+        Linear()
```
Comment (kouroshHakha) [inline rllib/core/models/configs.py:95]: what is this Linear()?

File: rllib/core/models/torch/primitives.py (hunk 6 above)
Comment (kouroshHakha) [reply rllib/core/models/torch/primitives.py:67]: ok makes sense.

File: rllib/core/models/torch/primitives.py (hunk 8 above)
Comment (kouroshHakha) [reply rllib/core/models/torch/primitives.py:108]: This should be commented.

**Review by kouroshHakha — APPROVED** [review]
only one nit remaining. We can merge as is.

File: rllib/core/models/configs.py (hunk 11 above)
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/configs.py]: Corrected, sorry.
